docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,404 @@
+    // matches this definition. This field is optional, and in case it is not set
+    // results in the rule being applied to all pods in the namespace.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+
+    // a collection of Egress NetworkQoS rule objects
+    Egress []Rule `json:"egress"`
+}
+
+type Rule struct {
+    // Priority The NetworkQoS rule’s priority. Rules with numerically higher
+    // priority take precedence over those with lower. If two NetworkQoS
+    // rules with the same priority both match, then the one
+    // actually applied to a packet is undefined.
+    // +kubebuilder:validation:Maximum:=32767
+    // +kubebuilder:validation:Minimum:=0
+    Priority int `json:"priority"`
+
+    // DSCP marking value for matching pods' traffic.
+    // +kubebuilder:validation:Maximum:=63
+    // +kubebuilder:validation:Minimum:=0
+    DSCP int `json:"dscp"`
+
+    // Classifer The classifier on which packets should match
+    // to apply the NetworkQoS Rule.
+    // This field is optional, and in case it is not set the rule is applied
+    // to all egress traffic regardless of the destination.
+    // +optional
+    Classifier Classifer `json:"classifier"`
+
+    // +optional
+    Bandwidth Bandwidth `json:"bandwidth"`
+}
+
+type Classifer struct {
+    // +optional
+    To []Destination `json:"to"`
+
+    // +optional
+    Ports Ports `json:"ports"`
+}
+
+// Bandwidth controls the maximum of rate traffic that can be sent
+// or received on the matching packets.
+type Bandwidth struct {
+    // Rate The value of rate limit in kbps. Traffic over the limit
+    // will be dropped.
+    // +kubebuilder:validation:Minimum:=1
+    // +kubebuilder:validation:Maximum:=4294967295
+    // +optional
+    Rate uint32 `json:"rate"`
+
+    // Burst The value of burst rate limit in kilobits.
+    // This also needs rate to be specified.
+    // +kubebuilder:validation:Minimum:=1
+    // +kubebuilder:validation:Maximum:=4294967295
+    // +optional
+    Burst uint32 `json:"burst"`
+}
+
+// Ports specifies destination protocol and port on which NetworkQoS
+// rule is applied
+type Ports struct {
+    // protocol (tcp, udp, sctp) that the traffic must match.
+    // +kubebuilder:validation:Pattern=^TCP|UDP|SCTP$
+    // +optional
+    Protocol string `json:"protocol"`
+
+    // port that the traffic must match
+    // +kubebuilder:validation:Minimum:=1
+    // +kubebuilder:validation:Maximum:=65535
+    // +optional
+    Port int32 `json:"port"`
+}
+
+// Destination specifies either pod or CIDR on which NetworkQoS rule is applied.
+type Destination struct {
+    // +optional
+    PodSelector PodSelector `json:"podSelector,omitempty"`
+
+    // CIDR specifies the destination's CIDR. Only traffic heading
+    // to this CIDR will be marked with the DSCP value and bandwidth
+    // get applied.
+    // +optional
+    // +kubebuilder:validation:Format="cidr"
+    CIDR *string `json:"cidr,omitempty"`
+}
```

Comment:
Can one select both? or they can only set one of the two?

Reply:
yes, with new changes, it can take both params.